### PR TITLE
fix: switch chain with walletconnect

### PR DESCRIPTION
**Pull request by ignaciosantise**
## **Description**
MetaMask calls updateSession multiple times when the user tries to switch chain from the dapp. This causes the dapp to receive a `chainChanged` event before receiving the updated session from the wallet.

* Compare addresses in lowercase
* Parse chainId correctly
* Clearing state variables of BackgroundBridge when the session is disconnected
* Bumped walletconnect sdk versions to latest

## **Related issues**

Fixes:

## **Manual testing steps**

1. Go to https://lab.web3modal.com/library/ethers/ with an iPhone
2. Press the "Connect Wallet" button and select "Metamask"
3. Press "Open" to open MetaMask (if applies)
4. Accept the session proposal
5. Go back to the browser
6. Switch chain to Polygon (chain must be Already added as know network in Metamask)
7. Go to MetaMask and accept the switch popup
8. Go back to the browser and see there is a new redirect popup, which causes issues in the Wallet. The popup shouldn't be there

### **Before**

https://github.com/user-attachments/assets/4a0b4a1c-6496-4e2c-8e03-d04a51df2ce3

### **After**

https://github.com/user-attachments/assets/7441400e-6dfe-4595-a608-2eec30b1c478

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [ ] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by ignaciosantise** [comment]
I have read the CLA Document and I hereby sign the CLA

File: app/core/BackgroundBridge/BackgroundBridge.js [hunk 1]
```diff
@@ -286,6 +286,18 @@ export class BackgroundBridge extends EventEmitter {
...
+      this.chainIdSent = publicState.chainId;
+    }
+
+    if (!this.networkVersionSent) {
+      this.networkVersionSent = publicState.networkVersion;
+    }
+
+    if (!this.addressSent) {
+      this.addressSent = memState.selectedAddress;
+    }
```
Comment (arein) [inline app/core/BackgroundBridge/BackgroundBridge.js:299]: why is this needed?

File: app/core/BackgroundBridge/BackgroundBridge.js [hunk 2]
```diff
@@ -296,10 +308,12 @@ export class BackgroundBridge extends EventEmitter {
...
       this.notifyChainChanged(publicState);
     }
-
     // ONLY NEEDED FOR WC FOR NOW, THE BROWSER HANDLES THIS NOTIFICATION BY ITSELF
     if (this.isWalletConnect || this.isRemoteConn) {
-      if (this.addressSent !== memState.selectedAddress) {
+      if (
+        this.addressSent?.toLowerCase() !==
+        memState.selectedAddress?.toLowerCase()
+      ) {
```
Comment (arein) [inline app/core/BackgroundBridge/BackgroundBridge.js:316]: is this needed somewhere else too?

File: app/core/BackgroundBridge/WalletConnectPort.ts [hunk 3]
```diff
@@ -31,7 +31,7 @@ class WalletConnectPort extends EventEmitter {
       } else if (msg?.data?.method === NOTIFICATION_NAMES.accountsChanged) {
         const chainId = selectChainId(store.getState());
         this._wcRequestActions?.updateSession?.({
-          chainId: parseInt(chainId, 10),
+          chainId: parseInt(chainId),
```
Comment (arein) [inline app/core/BackgroundBridge/WalletConnectPort.ts:34]: curious: why is this needed?

**Review by arein — APPROVED** [review]

File: app/core/BackgroundBridge/WalletConnectPort.ts (hunk 3 above)
Comment (ignaciosantise) [reply app/core/BackgroundBridge/WalletConnectPort.ts:34]: chainId is an hex value, so that parse won't work. 

`parseInt` default behavior is: 

`If the second argument is not supplied, strings with a prefix of '0x' are considered hexadecimal. All other strings are considered decimal.`

File: app/core/BackgroundBridge/BackgroundBridge.js (hunk 2 above)
Comment (ignaciosantise) [reply app/core/BackgroundBridge/BackgroundBridge.js:316]: 🤷‍♂️ not sure, maybe someone from MM team can check

**Review by andreahaku — DISMISSED** [review]
LGTM

**Comment by enesozturk** [comment]
Hey @andreahaku, what are the next steps to get the PRs merged and published? 🙏

**Comment by andreahaku** [comment]
> Hey @andreahaku, what are the next steps to get the PRs merged and published? 🙏

it needs to be QAd by our engineers. When it's done we can merge it. Thanks

**Comment by ignaciosantise** [comment]
@christopherferreira9 yes, to receive a new switch event on MM you need to send a request, like Personal Sign.

This is internally handled by the `se-sdk`.

**Comment by ganchoradkov** [comment]
> @ignaciosantise Is it normal that this only works on the first switching?
> 
>  Screen.Recording.2024-08-14.at.17.19.48.mov

Hey @christopherferreira9, this is the expected behaviour. The reason why you get prompt to accept the switch initially is because `polygon` (in this example) is not approved in the session. The wallet receives the request, updates the session with this new chain (polygon) and emits chainChanged event. The switch to ethereum is done only locally since it is already approved in the session. If you change again to polygon, it will be local change as well so you won't receive request in the wallet. If you change to 3rd chain though, you will again get request so the wallet can update the session.

Hope this context helps!

**Comment by christopherferreira9** [comment]
> > @ignaciosantise Is it normal that this only works on the first switching?
> > Screen.Recording.2024-08-14.at.17.19.48.mov
> 
> Hey @christopherferreira9, this is the expected behaviour. The reason why you get prompt to accept the switch initially is because `polygon` (in this example) is not approved in the session. The wallet receives the request, updates the session with this new chain (polygon) and emits chainChanged event. The switch to ethereum is done only locally since it is already approved in the session. If you change again to polygon, it will be local change as well so you won't receive request in the wallet. If you change to 3rd chain though, you will again get request so the wallet can update the session.
> 
> Hope this context helps!

Thank you for the reply. iOS and Android look good.
Tested with both deeplink and QR code session for both platforms.

**Review by andreahaku — APPROVED** [review]
LGTM

**Review by sethkfman — APPROVED** [review]
LGTM
